gazebo_ros/src/gazebo_ros_api_plugin.cpp

```diff
@@ -1940,6 +1945,50 @@ void GazeboRosApiPlugin::publishSimTime()
   pub_clock_.publish(ros_time_);
 }
 
+void GazeboRosApiPlugin::publishRealTimeFactor(ConstWorldStatisticsPtr &_msg)
+{
+  // Note that this is a reimplementation of the Gazebo method found here:
+  // https://bitbucket.org/osrf/gazebo/src/a24b331f8ebfd712a226ba73b7f43d2d4c14fe15/tools/gz.cc?at=gazebo7&fileviewer=file-view-default#gz.cc-854
+  // The goal is to make the real-time factor available in ROS bags, for use as an easy "system performance" metric
+  // with post-run analysis.
+  double percent = 0;
+  common::Time simTime  = msgs::Convert(_msg->sim_time());
+  common::Time realTime = msgs::Convert(_msg->real_time());
+
+  this->simTimes.push_back(msgs::Convert(_msg->sim_time()));
+  if (this->simTimes.size() > 20)
+    this->simTimes.pop_front();
+
+  this->realTimes.push_back(msgs::Convert(_msg->real_time()));
+  if (this->realTimes.size() > 20)
+    this->realTimes.pop_front();
+
+  common::Time simAvg, realAvg;
+  std::list<common::Time>::iterator simIter, realIter;
+  simIter = ++(this->simTimes.begin());
+  realIter = ++(this->realTimes.begin());
+  while (simIter != this->simTimes.end() && realIter != this->realTimes.end())
```

Comment:
Rather than recalculating the average every loop, you could store a running sum which you add the new time to and (if the buffer is full) pop and subtract the oldest time from.

Reply:
That gets a little redundant when using `boost::circular_buffer` - if we're tracking how full the buffer is anyway, we might as well stick with the `std::list`.

What I'm looking at doing instead is using `std::accumulate` to produce a sum every pass.  Looking at the [original implementation](https://bitbucket.org/osrf/gazebo/src/a24b331f8ebfd712a226ba73b7f43d2d4c14fe15/tools/gz.cc#lines-854:892), it's not _actually_ calculating an average anyway.